### ISSUE TITLE
Support Optax Optimizers (with JAX backend)

### ISSUE THREAD

**Issue by ixig** [keras-team/keras#18923]
It says [here](https://keras.io/guides/writing_a_custom_training_loop_in_jax/) that the use of Optax optimizer is supported in Keras3, but that's not currently the case. I would like to use Optax'es MultiSteps optimizer (wrapped around optax.Adam) to implement gradient accumulation over multiple training steps.

I believe being able to do this is very beneficial to many Keras users, as larger models are crimping batch sizes. This is very easy to do on PyTorch.

I managed to use the above guide to implement my own training loop, using the optax.MultiSteps -- but this naive custom loop leaves a lot to be desired -- i.e. it's not scalable to sharding (distributed training), losses all of the niceties of model.fit(), etc. . Can the Keras team please prioritize supporting the use of Optax optimizers (specifically [optax.MultiSteps](https://optax.readthedocs.io/en/latest/api.html?highlight=multisteps#optax.MultiSteps) wrapping optax.adam)?

P.S.: I tried to do this myself in the codebase, but got stuck at `JAXTrainer._record_training_state_sharding_spec()`.

**Comment by fchollet** [comment]
It sounds like what you need is simply gradient accumulation? It's pretty easy to implement in Keras, either via a custom optimizer (best option) or via a custom `train_step`. 

Optax optimizers are supported in Keras 3 in the sense that it's easy to write a JAX training loop to train a Keras 3 model using an Optax optimizer (same story with PyTorch optimizers). However `fit()` can only use Keras-native objects (whether optimizers or metrics).

Maybe we should consider implementing a `MultiSteps` equivalent in Keras?

**Comment by ixig** [comment]
@fchollet , yes certainly, if that's easier to add. PyTorch Lightning's Trainer supports this [out of the box](https://lightning.ai/docs/pytorch/stable/common/trainer.html#accumulate-grad-batches), and not to mention loss.backwards() by default accumulates gradients for this very reason. So yeah, either a MultiSteps equivalent optimizer-wrapper, or supporting this additional flag to Model.fit() would be much appreciated!

I was able to use the optax.MultiSteps optimizer in my version of a custom training loop using your guide (with slight changes for init and updating), but I fear it's not robust or scalable (I need it to work with Keras-JAX model-parallelism). Also tried implementing without it, but adhering to JAX'es i. purely functional ii. no python conditionals ... constraints was above my level of expertise.

I also took a crack at writing my own gradient accumulation optimizer in Keras, but ran into a roadblock because JAX'es JIT doesn't allow for python conditionals in-graph, which I was needing to gate the behavior to either accumulate gradients or to step the optimizer. JAX newbie here ... but isn't that the whole idea behind Keras3, it's JAX "for mere humans", LOL. Thanks!

**Comment by fchollet** [comment]
I think gradient accumulation is something we can add either as an optimizer wrapper, or as a base class argument. I guess optimizer wrapper may be best. Can you open a PR even if the code in the PR doesn't work? It could serve as a starting point, even if we don't end up merging it.

**Comment by ixig** [comment]
I've submitted my PR #18948 for this.

**Comment by sachinprasadhs** [comment]
@ixig , Could you please close the issue since it is addressed in the PR here https://github.com/keras-team/keras/pull/18951/files
